src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -341,170 +603,333 @@ function saturation_shum(T, ρ, q_liq=0, q_ice=0)
-(moist-)air density `ρ`, and total specific humidity `q_tot`.
+Partition the phases in equilibrium into
+ - the liquid specific humidity `q_liq`, and
+ - the ice specific humidity `q_ice`,
+using the [`liquid_fraction`](@ref) function given a thermodynamic state `ts`.
+"""
+phase_partitioning_eq(ts::ThermodynamicState) =
+  phase_partitioning_eq(air_temperature(ts), air_density(ts), ts.q_tot)
+phase_partitioning_eq(ts::PhaseNonEquil) = ts.q_liq, ts.q_ice
 
-The optional input value of the temperature `T_init` is taken as the initial
-value of the saturation adjustment iterations.
 """
-function saturation_adjustment(e_int, ρ, q_tot, T_init = T_triple)
+    saturation_adjustment(e_int, ρ, q_tot)
+
+Compute the temperature that is consistent with
+
+ - `e_int` internal energy
+ - `ρ` (moist-)air density
+ - `q_tot` total specific humidity
+
+See also [`saturation_adjustment_q_t_θ_l`](@ref).
+"""
+function saturation_adjustment(e_int::DT, ρ::DT, q_tot::DT) where DT
     if q_tot <= saturation_shum(max(0,air_temperature(e_int, q_tot)), ρ)
-      return air_temperature(e_int, q_tot)   
+      return air_temperature(e_int, q_tot)
    else
-    tol_abs = 1e-3*cv_d
+    tol_abs = DT(1e-3)*DT(cv_d)
     iter_max = 10
     args = (ρ, q_tot, e_int)
-    T0 = max(T_min, air_temperature(e_int, q_tot, 0.0, 0.0))
-    T1 = air_temperature(e_int, q_tot, 0.0, q_tot)
-    roots_equation(x, ρ, q_tot, e_int) = internal_energy_sat(x, ρ, q_tot) - e_int
-    T, converged = find_zero(roots_equation,
-                             T0, T1,
-                             args,
+    # FIXME here: need to revisit bounds for saturation adjustment to guarantee bracketing of zero.
+    T_1 = max(T_min, air_temperature(e_int, q_tot, DT(0), DT(0)))
+    T_2 = air_temperature(e_int, q_tot, DT(0), q_tot)
+    eos_root(x, ρ, q_tot, e_int) = internal_energy_sat(x, ρ, q_tot) - e_int
+    T, converged = find_zero(eos_root, T_1, T_2, args,
                              IterParams(tol_abs, iter_max),
-                             SecantMethod()
-                             )
+                             SecantMethod())
     q_liq, q_ice = phase_partitioning_eq(T, ρ, q_tot)
     return air_temperature(e_int, q_tot, q_liq, q_ice)
   end
 end
 
+
 """
-    liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
+    saturation_adjustment_q_tot_θ_liq_ice(θ_liq_ice, q_tot, ρ, p)
+
+Compute the temperature that is consistent with
+
+ - `θ_liq_ice` liquid-ice potential temperature
+ - `q_tot` total specific humidity
+ - `ρ` density
+ - `p` pressure
+
+See also [`saturation_adjustment`](@ref).
+"""
+function saturation_adjustment_q_tot_θ_liq_ice(θ_liq_ice::DT, q_tot::DT, ρ::DT, p::DT) where DT
+  T_1 = air_temperature_from_pottemp(θ_liq_ice, p) # Assume all vapor
+  qv_star = saturation_shum(T_1, ρ)
+  if (q_tot <= qv_star) # If not saturated
+    return T_1
+  else  # If saturated, iterate
+    T_2 = air_temperature_from_pottemp(θ_liq_ice, p, q_tot, DT(0), q_tot) # Assume all ice
```

Comment:
Same as above. Need to update the `air_temperature_from...` function so that this will work.